### PR TITLE
Authentication using JWT token

### PR DESCRIPTION
**Pull request by maxim12321**
Add custom UserDetails, storing the information about the user (now: login, password, role).
Configure Spring Security to use JWT token for authentication.

`/login` endpoint performs authentication using user's login and password, returns generated JWT token.
Other endpoints accept only authenticated users.

Provide unit-tests for services and JWT utilities.

### CODE REVIEW

File: src/main/java/com/team4/testingsystem/TestingSystemApplication.java [hunk 1]
```diff
@@ -5,7 +5,7 @@
 
 @SpringBootApplication
 public class TestingSystemApplication {
-	public static void main(String[] args) {
```
Comment (AntonMaksimenkau) [inline src/main/java/com/team4/testingsystem/TestingSystemApplication.java]: Please, use a single code style within your team

File: src/main/java/com/team4/testingsystem/config/WebSecurityConfiguration.java [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+import org.springframework.security.config.annotation.web.configuration.WebSecurityConfigurerAdapter;
+import org.springframework.security.config.http.SessionCreationPolicy;
+import org.springframework.security.crypto.password.NoOpPasswordEncoder;
+import org.springframework.security.crypto.password.PasswordEncoder;
+import org.springframework.security.web.authentication.UsernamePasswordAuthenticationFilter;
+import org.springframework.security.web.csrf.CookieCsrfTokenRepository;
+
+@EnableWebSecurity
+public class WebSecurityConfiguration extends WebSecurityConfigurerAdapter {
+    @Autowired
```
Comment (AntonMaksimenkau) [inline src/main/java/com/team4/testingsystem/config/WebSecurityConfiguration.java]: Currently, injection in a constructor known as a best practice. Let's use it

File: src/main/java/com/team4/testingsystem/config/WebSecurityConfiguration.java [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+                .anyRequest().authenticated()
+                .and()
+                .sessionManagement().sessionCreationPolicy(SessionCreationPolicy.STATELESS);
+
+        http.addFilterBefore(jwtRequestFilter, UsernamePasswordAuthenticationFilter.class);
+    }
+
+    @Bean
+    public PasswordEncoder getPasswordEncoder() {
+        return NoOpPasswordEncoder.getInstance();
```
Comment (AntonMaksimenkau) [inline src/main/java/com/team4/testingsystem/config/WebSecurityConfiguration.java]: NoopPasswordEncoder is insecure. Let's use BcryptPasswordEncoder instead

File: src/main/java/com/team4/testingsystem/controllers/AuthenticationController.java [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+import org.springframework.web.bind.annotation.RestController;
+
+import java.util.Optional;
+
+@RestController
+public class AuthenticationController {
+    @Autowired
+    private AuthenticationService authenticationService;
+
+    @GetMapping("/")
```
Comment (AntonMaksimenkau) [inline src/main/java/com/team4/testingsystem/controllers/AuthenticationController.java]: What is index page here and why do we need it?

**Review by AntonMaksimenkau — CHANGES_REQUESTED** [review]
Please, have a look at the comments

File: src/main/java/com/team4/testingsystem/controllers/AuthenticationController.java (hunk 4 above)
Comment (maxim12321) [reply src/main/java/com/team4/testingsystem/controllers/AuthenticationController.java]: It was used to make the first request and get CSRF token from the response
If the request fails (e.g. 403 error), will token still be attached to the responce header?

Comment (AntonMaksimenkau) [reply src/main/java/com/team4/testingsystem/controllers/AuthenticationController.java]: So it was used only for test purposes. Do you think we need it on a prod environment?

Comment (maxim12321) [reply src/main/java/com/team4/testingsystem/controllers/AuthenticationController.java]: 😳

**Review by AntonMaksimenkau — CHANGES_REQUESTED** [review]
Still have some issues with a code style. Please, set code style settings in your ide to default and make a refactoring

**Review by AntonMaksimenkau — DISMISSED** [review]

**Review by SemyonKorostik — DISMISSED** [review]

**Review by Darina007 — DISMISSED** [review]

File: src/main/java/com/team4/testingsystem/entities/UserRoleEntity.java [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+import javax.persistence.Table;
+import java.util.Collection;
+
+@Entity
+@Table(name = "user_role")
+public class UserRoleEntity {
+    @Id
+    @GeneratedValue
+    @Column(name = "id")
+    private Integer id;
```
Comment (AntonMaksimenkau) [inline src/main/java/com/team4/testingsystem/entities/UserRoleEntity.java]: Integer is too small for users.
Let's use at least Long

**Review by AntonMaksimenkau — CHANGES_REQUESTED** [review]

**Review by AntonMaksimenkau — APPROVED** [review]

**Review by Darina007 — DISMISSED** [review]

File: src/main/java/com/team4/testingsystem/entities/UserEntity.java [hunk 6]
```diff
@@ -0,0 +1,76 @@
+package com.team4.testingsystem.entities;
+
+import javax.persistence.*;
+
+@Entity
+@Table(name = "user")
+public class UserEntity {
```
Comment (AntonMaksimenkau) [inline src/main/java/com/team4/testingsystem/entities/UserEntity.java]: Do we really need Entity postifx in the entity package?

File: src/main/java/com/team4/testingsystem/entities/UserRoleEntity.java [hunk 7]
```diff
@@ -0,0 +1,43 @@
+package com.team4.testingsystem.entities;
+
+import javax.persistence.*;
+import java.util.Collection;
+
+@Entity
+@Table(name = "user_role")
+public class UserRoleEntity {
```
Comment (AntonMaksimenkau) [inline src/main/java/com/team4/testingsystem/entities/UserRoleEntity.java]: Same here

**Review by AntonMaksimenkau — CHANGES_REQUESTED** [review]

File: src/main/java/com/team4/testingsystem/entities/UserEntity.java (hunk 6 above)
Comment (maxim12321) [reply src/main/java/com/team4/testingsystem/entities/UserEntity.java]: Nope, not really)

**Review by AntonMaksimenkau — APPROVED** [review]

**Review by Anlord241 — APPROVED** [review]

**Pull request merged**